libraries/platforms/dromajo-vcs/bsg_manycore_platform.cpp

```diff
@@ -0,0 +1,491 @@
+  config_req_pkt.request.reg_id = 0;
+  // Note: The onus is on the simulator (i.e. x86 code) to check if the config
+  // index is within bounds before accessing the DPI ROM.
+  config_req_pkt.request.addr = HB_BP_HOST_EPA_CONFIG_START + (idx << 2);
+
+  // Note: Potentially dangerous to write to the FIFO without checking for credits
+  // We get back credits used and not credits remaining and without the configuration
+  // there is no way to know the credits remaining.
+  err = hb_bp_write_to_mc_bridge(&config_req_pkt);
+  if (err != HB_MC_SUCCESS) {
+    bsg_pr_err("Write to the host request FIFO failed!");
+    return err;
+  }
+
+  do {
+    err = hb_bp_get_fifo_entries(&num_entries, HB_MC_FIFO_RX_RSP);
+  } while ((num_entries == 0) || (err != HB_MC_SUCCESS));
+
+  err = hb_bp_read_from_mc_bridge(&config_resp_pkt, HB_MC_FIFO_RX_RSP);
+  if (err != HB_MC_SUCCESS) {
+    bsg_pr_err("Config read failed\n");
+    return HB_MC_FAIL;
+  }
+
+  *config = config_resp_pkt.response.data;
+
+  return HB_MC_SUCCESS;
+}
+
+/**
+ * Stall until the all requests (and responses) have reached their destination.
+ * @param[in]  mc     A manycore instance initialized with hb_mc_manycore_init()
+ * @param[in] timeout A timeout counter. Unused - set to -1 to wait forever.
+ * @return HB_MC_SUCCESS on success. Otherwise an error code defined in bsg_manycore_errno.h.
+ */
+int hb_mc_platform_fence(hb_mc_manycore_t *mc, long timeout) {
+  int credits_used, is_vacant, num_entries, err;
+  hb_mc_packet_t fence_req_pkt, fence_resp_pkt;
+  num_entries = 0;
+  credits_used = 0;
+
+  hb_mc_platform_t *platform = reinterpret_cast<hb_mc_platform_t *>(mc->platform); 
+
+  if (timeout != -1) {
+    manycore_pr_err(mc, "%s: Only a timeout value of -1 is supported\n", __func__);
+    return HB_MC_NOIMPL;
+  }
+
+  // Prepare host packet to query TX vacancy
+  fence_req_pkt.request.x_dst = HOST_X_COORD;
+  fence_req_pkt.request.y_dst = HOST_Y_COORD;
+  fence_req_pkt.request.x_src = BP_HOST_LINK_X;
+  fence_req_pkt.request.y_src = BP_HOST_LINK_Y;
+  fence_req_pkt.request.op_v2 = HB_MC_PACKET_OP_REMOTE_LOAD;
+  fence_req_pkt.request.payload = 0;
+  fence_req_pkt.request.reg_id = 0;
+  fence_req_pkt.request.addr = HB_BP_HOST_EPA_TX_VACANT; // x86 Host address to poll tx vacancy
+
+  do
+  {
+    err = hb_bp_get_credits_used(&credits_used);
+    // In a real system, this function call makes no sense since we will be sending packets to the
+    // host on the network and are trying to check for credits to be zero and complete the fence.
+    // It is fine here because of the system setup.
+    err |= hb_bp_write_to_mc_bridge(&fence_req_pkt);
+
+    do {
+      err = hb_bp_get_fifo_entries(&num_entries, HB_MC_FIFO_RX_RSP);
+    } while ((num_entries == 0) || (err != HB_MC_SUCCESS));
+
+    err |= hb_bp_read_from_mc_bridge(&fence_resp_pkt, HB_MC_FIFO_RX_RSP);
+    is_vacant = fence_resp_pkt.response.data;
+  } while ((err == HB_MC_SUCCESS) && !((credits_used == 0) && is_vacant));
```

Comment:
So BP counts down to 0, not up to the credit limit (in the manycore), is that correct?

Reply:
I would move error checking out of the while loop condition and make them discrete. It makes the termination condition hard to think about, and makes debugging later easier. (This would solve the |= comment above, too)

Then, I would make it: (credits_used != 0) || !is_vacant

This makes it clearer that the while loop will not terminate if either condition holds. Having the ! on the outside means I have to think about it and apply boolean algebra.

Reply:
Yes, BP counts down to 0 because the DPI module in the testbench returns the number of credits currently used (https://github.com/bespoke-silicon-group/bsg_manycore/blob/d7ab95c470bc3ade5dce2a18046339ea211f6a8c/testbenches/dpi/bsg_nonsynth_dpi_manycore.hpp#L116)